File: app/src/main/python/colormate/color_matcher.py
"""
ColorMatcher - 色彩风格匹配器

将训练好的色彩风格模型应用到目标照片上。
使用多层色彩迁移策略，从粗到细地调整目标照片的颜色。

迁移策略（按应用顺序）：
1. Lab色彩空间全局统计匹配（均值、标准差对齐）
2. 色调直方图匹配（Reinhard色彩迁移增强版）
3. 饱和度自适应映射
4. 色调映射曲线匹配（分位数匹配）
5. 局部对比度增强（可选）
6. 颜色自然度保护（防止色偏过重）
"""

import numpy as np
from scipy import interpolate
from scipy.ndimage import gaussian_filter

try:
    import cv2
    USE_CV2 = True
except ImportError:
    USE_CV2 = False
    from PIL import Image


class ColorMatcher:
    """色彩风格匹配器"""
    
    def __init__(self, strength=1.0, preserve_natural=True):
        """
        Args:
            strength: 迁移强度 (0.0 ~ 1.0)，1.0为完全迁移
            preserve_natural: 是否保护自然度（防止色偏）
        """
        self.strength = max(0.0, min(1.0, strength))
        self.preserve_natural = preserve_natural
    
    def match(self, model, target_path, output_path=None, strength=None):
        """
        对目标图片应用色彩风格迁移
        
        Args:
            model: ColorStyleModel 实例
            target_path: 目标图片路径
            output_path: 输出路径（可选，为None时返回数组）
            strength: 覆盖初始化时的强度设置
            
        Returns:
            如果 output_path 为 None，返回 RGB numpy array
            否则保存到文件
        """
        if strength is not None:
            self.strength = max(0.0, min(1.0, strength))
        
        s = self.strength
        
        # 读取目标图片
        img_bgr, img_rgb = self._read_image(target_path)
        if img_bgr is None:
            raise IOError(f"Cannot read image: {target_path}")
        
        original = img_rgb.copy().astype(np.float32)
        lab = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2LAB).astype(np.float32)
        
        # =============================================
        # 第1层：Lab统计匹配（基础色调调整）
        # =============================================
        lab_matched = self._statistical_match(
            lab, 
            np.array(model.lab_means),
            np.array(model.lab_stds),
            s
        )
        
        # =============================================
        # 第2层：色调映射曲线匹配（分位数匹配）
        # =============================================
        lab_curve = self._quantile_matching(
            lab_matched,
            np.array(model.l_quantiles),
            np.array(model.a_quantiles),
            np.array(model.b_quantiles),
            s * 0.8  # 曲线匹配强度稍低
        )
        
        # =============================================
        # 第3层：色调直方图匹配
        # =============================================
        # 先转回RGB进行处理
        rgb_mid = cv2.cvtColor(lab_curve.astype(np.uint8), cv2.COLOR_LAB2RGB).astype(np.float32)
        
        # 色调匹配在HSV空间做
        hsv = cv2.cvtColor(rgb_mid.astype(np.uint8), cv2.COLOR_RGB2HSV).astype(np.float32)
        target_hue_hist = np.array(model.hue_hist)
        hsv_matched = self._hue_matching(hsv, target_hue_hist, s * 0.5)
        
        # 饱和度匹配
        sat_stats = model.saturation_stats
        hsv_matched = self._saturation_matching(hsv_matched, sat_stats, s * 0.6)
        
        # 转换回RGB
        rgb_mid2 = cv2.cvtColor(hsv_matched.astype(np.uint8), cv2.COLOR_HSV2RGB).astype(np.float32)
        
        # =============================================
        # 第4层：自然度保护
        # =============================================
        if self.preserve_natural:
            result = self._blend_with_original(original, rgb_mid2, s)
        else:
            result = rgb_mid2
        
        # 剪裁到有效范围
        result = np.clip(result, 0, 255).astype(np.uint8)
        
        if output_path:
            result_bgr = cv2.cvtColor(result, cv2.COLOR_RGB2BGR)
            cv2.imwrite(str(output_path), result_bgr)
            return str(output_path)
        
        return result
    
    def _statistical_match(self, lab, target_means, target_stds, strength):
        """
        Lab空间统计匹配：均值对齐 + 标准差缩放
        
        这是Reinhard色彩迁移的核心算法，带保护机制
        """
        lab_out = lab.copy()
        
        # 分离通道
        l = lab[:, :, 0]
        a = lab[:, :, 1]
        b = lab[:, :, 2]
        
        # 均值对齐
        l_mean, l_std = np.mean(l), np.std(l)
        a_mean, a_std = np.mean(a), np.std(a)
        b_mean, b_std = np.mean(b), np.std(b)
        
        # Lab L范围[0,255], a/b范围[-128,127]
        # 保护边缘值不被过冲
        
        # L通道调整
        if l_std > 0:
            l_new = (l - l_mean) * (target_stds[0] / l_std) + target_means[0]
            l_new = l * (1 - strength) + l_new * strength
            # L通道有效范围 [0, 255]
            l_new = np.clip(l_new, 0, 255)
            lab_out[:, :, 0] = l_new
        
        # a通道调整
        if a_std > 0:
            a_new = (a - a_mean) * (target_stds[1] / a_std) + target_means[1]
            a_new = a * (1 - strength) + a_new * strength
            lab_out[:, :, 1] = a_new
        
        # b通道调整
        if b_std > 0:
            b_new = (b - b_mean) * (target_stds[2] / b_std) + target_means[2]
            b_new = b * (1 - strength) + b_new * strength
            lab_out[:, :, 2] = b_new
        
        return lab_out
    
    def _quantile_matching(self, lab, target_l_q, target_a_q, target_b_q, strength):
        """
        分位数匹配：将目标图片的每个Lab通道的亮度分布映射到参考风格的分位数曲线
        
        这样可以精确控制影调分布
        """
        lab_out = lab.copy().astype(np.float32)
        n_quantiles = len(target_l_q)
        quantiles = np.linspace(0, 1, n_quantiles)
        
        # 为每个通道创建映射函数
        for ch_idx, target_q in enumerate([target_l_q, target_a_q, target_b_q]):
            channel = lab[:, :, ch_idx].flatten().astype(np.float32)
            
            # 计算当前图片的分位数
            src_q = np.quantile(channel, quantiles)
            
            # 创建差值映射
            # 对每个像素: 找到其在源分位数中的位置，映射到目标分位数
            # 使用线性插值
            f = interpolate.interp1d(
                src_q, target_q, 
                bounds_error=False, 
                fill_value=(target_q[0], target_q[-1])
            )
            
            mapped = f(channel)
            
            # 混合
            blended = channel * (1 - strength) + mapped * strength
            lab_out[:, :, ch_idx] = blended.reshape(lab.shape[:2])
        
        return lab_out
    
    def _hue_matching(self, hsv, target_hue_hist, strength):
        """
        色调直方图匹配：调整HSV空间的H通道，使色调分布接近参考风格
        
        使用直方图规定化（Histogram Specification）
        """
        if strength <= 0:
            return hsv
        
        hsv_out = hsv.copy()
        h_channel = hsv[:, :, 0]  # H: 0-180 (OpenCV)
        
        # 目标色调直方图（从360 bin重采样到180 bin）
        target_h_180 = np.array(target_hue_hist)
        if len(target_h_180) == 360:
            target_h_180 = target_h_180.reshape(2, 180).mean(axis=0)
        
        # 计算当前图片的H通道直方图
        h_flat = h_channel.flatten()
        hist_curr, _ = np.histogram(h_flat, bins=180, range=(0, 180))
        hist_curr = hist_curr.astype(np.float32)
        hist_curr = hist_curr / max(hist_curr.sum(), 1)
        
        # 直方图规定化
        # 计算累积分布
        cdf_curr = np.cumsum(hist_curr)
        cdf_curr = cdf_curr / cdf_curr[-1]
        
        cdf_target = np.cumsum(target_h_180)
        cdf_target = cdf_target / cdf_target[-1]
        
        # 建立映射: 源bins -> 目标bins
        mapping = np.zeros(180, dtype=np.float32)
        for i in range(180):
            diff = np.abs(cdf_curr[i] - cdf_target)
            mapping[i] = np.argmin(diff)
        
        # 应用映射（带羽化避免Blocky artifact）
        mapped_h = mapping[h_flat.astype(np.int32)]
        
        # 混合
        h_new = h_flat * (1 - strength) + mapped_h * strength
        h_new = np.clip(h_new, 0, 179)
        
        hsv_out[:, :, 0] = h_new.reshape(h_channel.shape)
        
        return hsv_out
    
    def _saturation_matching(self, hsv, target_sat_stats, strength):
        """
        饱和度匹配：调整S通道的均值和标准差
        """
        if strength <= 0:
            return hsv
        
        hsv_out = hsv.copy()
        s_channel = hsv[:, :, 1].astype(np.float32)
        
        s_mean = np.mean(s_channel)
        s_std = np.std(s_channel)
        
        t_mean, t_std = target_sat_stats
        
        # 保护低饱和度区域（如天空、皮肤）
        if s_std > 0:
            # 计算每个像素的调整量，与当前饱和度值成正比
            s_new = (s_channel - s_mean) * (t_std / max(s_std, 1.0)) + t_mean
            
            # 对原饱和度很低的区域施加更少的调整
            low_sat_mask = s_channel < 30
            blend_factor = np.where(low_sat_mask, 0.3, 1.0)
            
            s_blended = s_channel * (1 - strength * blend_factor) + s_new * strength * blend_factor
            s_blended = np.clip(s_blended, 0, 255)
            
            hsv_out[:, :, 1] = s_blended
        
        return hsv_out
    
    def _blend_with_original(self, original, matched, strength):
        """
        与原图混合，保留部分原始细节和自然感
        
        同时使用边缘保护蒙版：在边缘区域保持更多原始信息
        """
        original_f = original.astype(np.float32)
        matched_f = matched.astype(np.float32)
        
        # 计算Lab空间的颜色偏移幅度
        orig_bgr = cv2.cvtColor(original_f.astype(np.uint8), cv2.COLOR_RGB2BGR).astype(np.float32)
        match_bgr = cv2.cvtColor(matched_f.astype(np.uint8), cv2.COLOR_RGB2BGR).astype(np.float32)
        
        orig_lab = cv2.cvtColor(orig_bgr.astype(np.uint8), cv2.COLOR_BGR2LAB).astype(np.float32)
        match_lab = cv2.cvtColor(match_bgr.astype(np.uint8), cv2.COLOR_BGR2LAB).astype(np.float32)
        
        # ab平面上的色偏幅度
        ab_diff = np.sqrt(
            (match_lab[:, :, 1] - orig_lab[:, :, 1]) ** 2 + 
            (match_lab[:, :, 2] - orig_lab[:, :, 2]) ** 2
        )
        
        # 色偏保护：色偏大的区域，用更低的强度混合
        max_diff = np.percentile(ab_diff, 95)
        if max_diff > 0:
            protect_mask = np.clip(1.0 - ab_diff / max_diff, 0.2, 1.0)
        else:
            protect_mask = np.ones_like(ab_diff)
        
        # 边缘检测：边缘区域保留更多原始细节
        gray = cv2.cvtColor(matched_f.astype(np.uint8), cv2.COLOR_RGB2GRAY)
        edges = cv2.Canny(gray, 50, 150).astype(np.float32) / 255.0
        edges = gaussian_filter(edges, sigma=2.0)
        
        # 在边缘区域，保留更多原始信息
        edge_protect = 1.0 - edges * 0.3  # 边缘区域最多保留30%原始
        
        # 合并保护蒙版
        combined_mask = protect_mask * edge_protect
        
        # 应用混合
        # 低强度时更保守，高强度时更大胆
        alpha = strength * combined_mask
        alpha = np.clip(alpha, 0, 1)
        
        # 3通道蒙版
        mask_3ch = np.stack([alpha, alpha, alpha], axis=2)
        
        result = original_f * (1 - mask_3ch) + matched_f * mask_3ch
        
        # 额外保护：如果ab偏移过大，退回到更自然的版本
        extreme_mask = (match_lab[:, :, 1] - orig_lab[:, :, 1]) > 50
        extreme_mask |= (match_lab[:, :, 2] - orig_lab[:, :, 2]) > 50
        if np.any(extreme_mask):
            extreme_mask_3ch = np.stack([extreme_mask] * 3, axis=2)
            result = np.where(
                extreme_mask_3ch,
                original_f * 0.5 + matched_f * 0.5,
                result
            )
        
        return result
    
    def _read_image(self, img_path):
        """读取图片"""
        path = str(img_path)
        if USE_CV2:
            img = cv2.imread(path, cv2.IMREAD_COLOR)
            if img is None:
                from PIL import Image
                pil_img = Image.open(path).convert('RGB')
                img_rgb = np.array(pil_img)
                img_bgr = cv2.cvtColor(img_rgb, cv2.COLOR_RGB2BGR)
                return img_bgr, img_rgb
            img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
            return img, img_rgb
        else:
            from PIL import Image
            pil_img = Image.open(path).convert('RGB')
            img_rgb = np.array(pil_img)
            img_bgr = cv2.cvtColor(img_rgb, cv2.COLOR_RGB2BGR)
            return img_bgr, img_rgb

File: app/src/main/java/com/colormate/app/color/ColorMatcher.kt
package com.colormate.app.color

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import org.opencv.android.Utils
import org.opencv.core.*
import org.opencv.imgproc.Imgproc
import kotlin.math.sqrt
import kotlin.math.min
import kotlin.math.max
import kotlin.math.ceil
import java.io.File

/**
 * 色彩风格匹配器 — 将训练好的色彩风格模型应用到目标照片上
 *
 * 对应原Python版 ColorMatcher
 *
 * 迁移策略（与原Python版一致）：
 * 1. Lab色彩空间全局统计匹配（均值、标准差对齐）
 * 2. 色调直方图匹配（Reinhard色彩迁移增强版）
 * 3. 饱和度自适应映射
 * 4. 色调映射曲线匹配（分位数匹配）
 * 5. 颜色自然度保护（防止色偏过重）
 */
class ColorMatcher(
    private var strength: Float = 1.0f,
    private val preserveNatural: Boolean = true
) {

    companion object {
        private const val HUE_BINS = 360
        private const val QUANTILE_N = 100
    }

    /**
     * 对目标图片应用色彩风格迁移
     * @param model 训练好的 ColorStyleModel
     * @param targetPath 目标图片路径
     * @param outputPath 输出路径
     * @param matchStrength 可选的强度覆盖
     */
    fun match(
        model: ColorStyleModel,
        targetPath: String,
        outputPath: String,
        matchStrength: Float? = null
    ) {
        if (matchStrength != null) strength = matchStrength
        strength = strength.coerceIn(0f, 1f)
        val s = strength

        // 读取目标图片为 Bitmap → Mat
        val options = BitmapFactory.Options().apply { inPreferredConfig = Bitmap.Config.ARGB_8888 }
        val bitmap = BitmapFactory.decodeFile(targetPath, options)
            ?: throw java.io.IOException("Cannot read image: $targetPath")

        val originalRgb = Mat()
        Utils.bitmapToMat(bitmap, originalRgb)

        // 保存原始RGB用于后续混合
        val original = Mat()
        originalRgb.copyTo(original)

        // 转Lab
        val lab = Mat()
        Imgproc.cvtColor(originalRgb, lab, Imgproc.COLOR_RGB2Lab)
        val labFloat = Mat()
        lab.convertTo(labFloat, CvType.CV_32FC3)

        // ========== 第1层：Lab统计匹配 ==========
        val labStats = statisticalMatch(labFloat, model.labMeans!!, model.labStds!!, s)

        // ========== 第2层：色调映射曲线匹配 ==========
        val labCurve = quantileMatching(labStats, model.lQuantiles!!, model.aQuantiles!!, model.bQuantiles!!, s * 0.8f)

        // ========== 第3层：色调+饱和度匹配 ==========
        val labCurveU8 = Mat()
        labCurve.convertTo(labCurveU8, CvType.CV_8UC3)
        val rgbMid = Mat()
        Imgproc.cvtColor(labCurveU8, rgbMid, Imgproc.COLOR_Lab2RGB)

        val hsv = Mat()
        Imgproc.cvtColor(rgbMid, hsv, Imgproc.COLOR_RGB2HSV)
        val hsvFloat = Mat()
        hsv.convertTo(hsvFloat, CvType.CV_32FC3)

        // 色调匹配
        val hsvHue = hueMatching(hsvFloat, model.hueHist!!, s * 0.5f)

        // 饱和度匹配
        val hsvSat = saturationMatching(hsvHue, model.saturationStats!!, s * 0.6f)

        val hsvSatU8 = Mat()
        hsvSat.convertTo(hsvSatU8, CvType.CV_8UC3)
        val rgbMid2 = Mat()
        Imgproc.cvtColor(hsvSatU8, rgbMid2, Imgproc.COLOR_HSV2RGB)

        // ========== 第4层：自然度保护 ==========
        val result: Mat
        if (preserveNatural) {
            result = blendWithOriginal(original, rgbMid2, s)
        } else {
            result = rgbMid2
        }

        // 剪裁并保存
        val resultClipped = Mat()
        Core.max(result, Scalar.all(0.0), resultClipped)
        Core.min(resultClipped, Scalar.all(255.0), resultClipped)
        resultClipped.convertTo(resultClipped, CvType.CV_8UC3)

        // 转BGR保存（OpenCV imwrite需要BGR）
        val resultBgr = Mat()
        Imgproc.cvtColor(resultClipped, resultBgr, Imgproc.COLOR_RGB2BGR)

        val outBitmap = Bitmap.createBitmap(resultBgr.cols(), resultBgr.rows(), Bitmap.Config.ARGB_8888)
        Utils.matToBitmap(resultBgr, outBitmap)

        File(outputPath).parentFile?.mkdirs()
        File(outputPath).outputStream().use { fos ->
            outBitmap.compress(Bitmap.CompressFormat.JPEG, 95, fos)
        }
    }

    // ========== 各层匹配算法 ==========

    /**
     * Lab空间统计匹配：均值对齐 + 标准差缩放
     * 对应Python版 _statistical_match
     */
    private fun statisticalMatch(
        lab: Mat, targetMeans: FloatArray, targetStds: FloatArray, s: Float
    ): Mat {
        val result = Mat()
        lab.copyTo(result)

        val channels = mutableListOf<Mat>()
        Core.split(result, channels)

        for (ch in 0..2) {
            val channel = channels[ch]
            val mean = Core.mean(channel).`val`[0]
            val std = computeStdDev(channel, mean)

            if (std > 0) {
                val newChannel = Mat()
                Core.subtract(channel, Scalar.all(mean.toDouble()), newChannel)
                Core.multiply(newChannel, Scalar.all((targetStds[ch] / std).toDouble()), newChannel)
                Core.add(newChannel, Scalar.all(targetMeans[ch].toDouble()), newChannel)

                // 混合
                Core.addWeighted(channel, (1 - s).toDouble(), newChannel, s.toDouble(), 0.0, channel)
            }

            // L通道 [0, 255], a/b范围约[-128, 127]
            when (ch) {
                0 -> Core.max(Scalar.all(0.0), channel, channel).also {
                    Core.min(channel, Scalar.all(255.0), channel)
                }
                1, 2 -> {
                    Core.max(Scalar.all(-128.0), channel, channel).also {
                        Core.min(channel, Scalar.all(127.0), channel)
                    }
                }
            }
        }

        Core.merge(channels, result)
        return result
    }

    /**
     * 分位数匹配：将目标图片的Lab通道分布映射到参考风格的分位数曲线
     * 对应Python版 _quantile_matching
     */
    private fun quantileMatching(
        lab: Mat, targetLQ: FloatArray, targetAQ: FloatArray, targetBQ: FloatArray, s: Float
    ): Mat {
        val result = Mat()
        lab.copyTo(result)

        val channels = mutableListOf<Mat>()
        Core.split(result, channels)

        val targetQs = listOf(targetLQ, targetAQ, targetBQ)

        for (ch in 0..2) {
            val channel = channels[ch]
            val totalPixels = channel.rows() * channel.cols()
            val flat = FloatArray(totalPixels)
            channel.get(0, 0, flat)

            // 计算当前图片分位数
            val sorted = flat.sorted().toFloatArray()
            val srcQ = computeQuantiles(sorted, QUANTILE_N)

            // 线性插值映射
            val mapped = FloatArray(flat.size)
            for (i in flat.indices) {
                mapped[i] = interpolate1D(srcQ, targetQs[ch], flat[i])
            }

            // 混合
            for (i in flat.indices) {
                flat[i] = flat[i] * (1 - s) + mapped[i] * s
            }
            channel.put(0, 0, flat)
        }

        Core.merge(channels, result)
        return result
    }

    /**
     * 色调直方图匹配：调整HSV空间的H通道
     * 对应Python版 _hue_matching
     */
    private fun hueMatching(hsv: Mat, targetHueHist: FloatArray, s: Float): Mat {
        if (s <= 0) return hsv

        val result = Mat()
        hsv.copyTo(result)

        val channels = mutableListOf<Mat>()
        Core.split(result, channels)
        val hChannel = channels[0]

        val totalPixels = hChannel.rows() * hChannel.cols()
        val flat = FloatArray(totalPixels)
        hChannel.get(0, 0, flat)

        // 目标直方图（360 bin → 180 bin for OpenCV）
        val targetH180 = if (targetHueHist.size == 360) {
            FloatArray(180) { i -> (targetHueHist[i * 2] + targetHueHist[i * 2 + 1]) / 2f }
        } else {
            targetHueHist
        }

        // 当前直方图 (180 bins, H: 0-180)
        val histCurr = FloatArray(180)
        for (v in flat) {
            val bin = (v / 180f * 180).toInt().coerceIn(0, 179)
            histCurr[bin]++
        }
        val histSum = histCurr.sum()
        if (histSum > 0) for (i in histCurr.indices) histCurr[i] /= histSum

        // CDF计算
        val cdfCurr = computeCdf(histCurr)
        val cdfTarget = computeCdf(targetH180)

        // 建立映射: 源bin → 目标bin
        val mapping = IntArray(180)
        for (i in 0 until 180) {
            var bestJ = 0
            var bestDiff = Float.MAX_VALUE
            for (j in 0 until 180) {
                val diff = kotlin.math.abs(cdfCurr[i] - cdfTarget[j])
                if (diff < bestDiff) {
                    bestDiff = diff
                    bestJ = j
                }
            }
            mapping[i] = bestJ
        }

        // 应用映射并混合
        for (i in flat.indices) {
            val srcBin = (flat[i] / 180f * 180).toInt().coerceIn(0, 179)
            val mappedValue = mapping[srcBin].toFloat() * 180f / 180f
            flat[i] = flat[i] * (1 - s) + mappedValue * s
            flat[i] = flat[i].coerceIn(0f, 179f)
        }
        hChannel.put(0, 0, flat)

        Core.merge(channels, result)
        return result
    }

    /**
     * 饱和度匹配：调整S通道的均值和标准差
     * 对应Python版 _saturation_matching
     */
    private fun saturationMatching(hsv: Mat, targetSatStats: FloatArray, s: Float): Mat {
        if (s <= 0) return hsv

        val result = Mat()
        hsv.copyTo(result)

        val channels = mutableListOf<Mat>()
        Core.split(result, channels)
        val sChannel = channels[1]

        val totalPixels = sChannel.rows() * sChannel.cols()
        val flat = FloatArray(totalPixels)
        sChannel.get(0, 0, flat)

        val sMean = flat.average().toFloat()
        val sStd = computeStdDev(flat, sMean)

        val tMean = targetSatStats[0]
        val tStd = targetSatStats[1]

        if (sStd > 0) {
            for (i in flat.indices) {
                val newVal = (flat[i] - sMean) * (tStd / sStd) + tMean

                // 低饱和度区域保护
                val blendFactor = if (flat[i] < 30f) 0.3f else 1.0f
                val blended = flat[i] * (1 - s * blendFactor) + newVal * s * blendFactor
                flat[i] = blended.coerceIn(0f, 255f)
            }
        }
        sChannel.put(0, 0, flat)

        Core.merge(channels, result)
        return result
    }

    /**
     * 与原图混合，保留原始细节和自然感
     * 对应Python版 _blend_with_original
     */
    private fun blendWithOriginal(original: Mat, matched: Mat, s: Float): Mat {
        if (original.size() != matched.size()) {
            val resized = Mat()
            Imgproc.resize(matched, resized, original.size())
            return blendWithOriginal(original, resized, s)
        }

        val origF = Mat()
        original.convertTo(origF, CvType.CV_32FC3)

        val matchF = Mat()
        matched.convertTo(matchF, CvType.CV_32FC3)

        // 转Lab计算色偏
        val origLab = Mat()
        val matchLab = Mat()
        Imgproc.cvtColor(original, origLab, Imgproc.COLOR_RGB2Lab)
        Imgproc.cvtColor(matched, matchLab, Imgproc.COLOR_RGB2Lab)

        val origLabF = Mat()
        val matchLabF = Mat()
        origLab.convertTo(origLabF, CvType.CV_32FC3)
        matchLab.convertTo(matchLabF, CvType.CV_32FC3)

        val totalPixels = origLabF.rows() * origLabF.cols()
        val origAB = FloatArray(totalPixels * 2)  // [a1, b1, a2, b2, ...]
        val matchAB = FloatArray(totalPixels * 2)

        // Extract a, b channels
        val origChannels = mutableListOf<Mat>()
        val matchChannels = mutableListOf<Mat>()
        Core.split(origLabF, origChannels)
        Core.split(matchLabF, matchChannels)

        val origA = FloatArray(totalPixels)
        val origB = FloatArray(totalPixels)
        val matchA = FloatArray(totalPixels)
        val matchB = FloatArray(totalPixels)
        origChannels[1].get(0, 0, origA)
        origChannels[2].get(0, 0, origB)
        matchChannels[1].get(0, 0, matchA)
        matchChannels[2].get(0, 0, matchB)

        // 计算ab色偏幅度
        val abDiff = FloatArray(totalPixels)
        var maxDiff = 0f
        for (i in 0 until totalPixels) {
            val d = sqrt(
                (matchA[i] - origA[i]) * (matchA[i] - origA[i]) +
                (matchB[i] - origB[i]) * (matchB[i] - origB[i])
            )
            abDiff[i] = d
            if (d > maxDiff) maxDiff = d
        }

        // 保护蒙版
        val protectMask = if (maxDiff > 0) {
            FloatArray(totalPixels) { (1f - abDiff[it] / maxDiff).coerceIn(0.2f, 1f) }
        } else {
            FloatArray(totalPixels) { 1f }
        }

        // 边缘检测保护
        val gray = Mat()
        Imgproc.cvtColor(matched, gray, Imgproc.COLOR_RGB2GRAY)
        val edges = Mat()
        Imgproc.Canny(gray, edges, 50.0, 150.0)

        val edgeFlat = ByteArray(totalPixels)
        edges.get(0, 0, edgeFlat)

        // 应用混合
        val result = Mat(origF.size(), CvType.CV_32FC3)
        val origFlat = FloatArray(totalPixels * 3)
        val matchFlat = FloatArray(totalPixels * 3)
        val resultFlat = FloatArray(totalPixels * 3)
        origF.get(0, 0, origFlat)
        matchF.get(0, 0, matchFlat)

        for (i in 0 until totalPixels) {
            val edgeFactor = 1.0f - (edgeFlat[i].toInt() and 0xFF) / 255f * 0.3f
            val alpha = (s * protectMask[i] * edgeFactor).coerceIn(0f, 1f)

            for (ch in 0..2) {
                resultFlat[i * 3 + ch] = origFlat[i * 3 + ch] * (1 - alpha) +
                        matchFlat[i * 3 + ch] * alpha
            }
        }
        result.put(0, 0, resultFlat)

        return result
    }

    // ========== 工具方法 ==========

    private fun computeStdDev(mat: Mat, mean: Double): Float {
        val flat = FloatArray(mat.rows() * mat.cols())
        mat.get(0, 0, flat)
        return computeStdDev(flat, mean.toFloat())
    }

    private fun computeStdDev(values: FloatArray, mean: Float): Float {
        if (values.isEmpty()) return 0f
        val variance = values.map { (it - mean) * (it - mean) }.average()
        return sqrt(variance).toFloat()
    }

    private fun computeQuantiles(sorted: FloatArray, n: Int): FloatArray {
        val result = FloatArray(n)
        for (i in 0 until n) {
            val q = i.toFloat() / (n - 1)
            val idx = q * (sorted.size - 1)
            val idxLow = idx.toInt().coerceIn(0, sorted.size - 1)
            val idxHigh = (idxLow + 1).coerceAtMost(sorted.size - 1)
            val frac = idx - idxLow
            result[i] = sorted[idxLow] * (1 - frac) + sorted[idxHigh] * frac
        }
        return result
    }

    /**
     * 一维线性插值
     */
    private fun interpolate1D(x: FloatArray, y: FloatArray, query: Float): Float {
        if (query <= x.first()) return y.first()
        if (query >= x.last()) return y.last()

        var low = 0
        var high = x.size - 1
        while (high - low > 1) {
            val mid = (low + high) / 2
            if (x[mid] <= query) low = mid
            else high = mid
        }

        val frac = (query - x[low]) / (x[high] - x[low])
        return y[low] * (1 - frac) + y[high] * frac
    }

    private fun computeCdf(hist: FloatArray): FloatArray {
        val cdf = FloatArray(hist.size)
        var sum = 0f
        for (i in hist.indices) {
            sum += hist[i]
            cdf[i] = sum
        }
        if (cdf.last() > 0) {
            for (i in cdf.indices) cdf[i] /= cdf.last()
        }
        return cdf
    }
}
